include/alpaka/warp/Traits.hpp

```diff
@@ -182,19 +194,115 @@ namespace alpaka::warp
     //! \param  width   number of threads receiving a single value
     //! \return val from the thread index srcLane.
     ALPAKA_NO_HOST_ACC_WARNING
-    template<typename TWarp>
-    ALPAKA_FN_ACC auto shfl(TWarp const& warp, std::int32_t value, std::int32_t srcLane, std::int32_t width = 0)
+    template<typename TWarp, typename T>
+    ALPAKA_FN_ACC auto shfl(TWarp const& warp, T value, std::int32_t srcLane, std::int32_t width = 0)
     {
         using ImplementationBase = concepts::ImplementationBase<ConceptWarp, TWarp>;
         return trait::Shfl<ImplementationBase>::shfl(warp, value, srcLane, width ? width : getSize(warp));
     }
 
-    //! shfl for float vals
+    //! Exchange data between threads within a warp.
+    //! It copies from a lane with lower ID relative to caller.
+    //! The lane ID is calculated by subtracting delta from the caller’s lane ID.
+    //!
+    //! Effectively executes:
+    //!
+    //!     __shared__ int32_t values[warpsize];
+    //!     values[threadIdx.x] = value;
+    //!     __syncthreads();
+    //!     return values[(-delta + width*floor(threadIdx.x/width))%width];
```

Comment:
This formula seems wrong:
 - `threadIdx.x/width` will be an integer between `0` and `blockDim.x / width`
 - passing it through `floor(...)` is unnecessary, and will be the same value
 - `width * floor(threadIdx.x/width)` will be a multiple of `width`
 - `(-delta + width*floor(threadIdx.x/width))` or `(width*floor(threadIdx.x/width) - delta)` will be a multiple of `width`, minus `delta`
 - then, `(...) % width` is equivalent to `(width - delta) % width`

which is the same value independently of `threadIdx.x`.